backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/tournament/TournamentService.java

```diff
@@ -0,0 +1,343 @@
+        User user = checkUser(userId);
+        Tournament tournament = checkTournament(tournamentDto.getId());
+
+        if (!tournament.getParticipants().contains(user)) {
+            throw new TutorException(USER_NOT_JOINED, user.getUsername());
+        }
+
+        tournament.removeParticipant(user);
+    }
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public TournamentDto updateTournament(Integer userId, Set<Integer> topicsId, TournamentDto tournamentDto) {
+        User user = checkUser(userId);
+        Tournament tournament = checkTournament(tournamentDto.getId());
+
+        if (tournament.getCreator() != user) {
+            throw new TutorException(TOURNAMENT_CREATOR, user.getId());
+        }
+
+        tournament.checkCanChange();
+
+        // change
+        if (DateHandler.isValidDateFormat(tournamentDto.getStartTime())) {
+            DateHandler.toISOString(tournament.getStartTime());
+            tournament.setStartTime(DateHandler.toLocalDateTime(tournamentDto.getStartTime()));
+        }
+
+        if (DateHandler.isValidDateFormat(tournamentDto.getEndTime())) {
+            DateHandler.toISOString(tournament.getEndTime());
+            tournament.setEndTime(DateHandler.toLocalDateTime(tournamentDto.getEndTime()));
+        }
+
+        if (tournament.hasQuiz()) { // update current Quiz
+            Quiz quiz = quizRepository.findById(tournamentDto.getQuizId())
+                    .orElseThrow(() -> new TutorException(QUIZ_NOT_FOUND, tournamentDto.getQuizId()));
+
+            QuizDto quizDto = quizService.findById(tournamentDto.getQuizId());
+            quizDto.setNumberOfQuestions(tournamentDto.getNumberOfQuestions());
+
+            quizService.updateQuiz(quiz.getId(), quizDto);
+        }
+        tournament.setNumberOfQuestions(tournamentDto.getNumberOfQuestions());
+
+        Set<Topic> topics = new HashSet<>();
+        for (Integer topicId : topicsId) {
+            Topic topic = topicRepository.findById(topicId)
+                    .orElseThrow(() -> new TutorException(TOPIC_NOT_FOUND, topicId));
+            tournament.checkTopicCourse(topic);
+            topics.add(topic);
+        }
+
+        tournament.updateTopics(topics);
+        return new TournamentDto(tournament);
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void cancelTournament(Integer userId, TournamentDto tournamentDto) {
+        User user = checkUser(userId);
+        Tournament tournament = checkTournament(tournamentDto.getId());
+
+        if (tournament.getCreator() != user) {
+            throw new TutorException(TOURNAMENT_CREATOR, user.getId());
+        }
+
+        tournament.checkCanChange();
+        tournament.setState(Tournament.Status.CANCELED);
```

Comment:
It would be nice the inform participants, a mailer service will be available soon